### PR TITLE
Economic indicators

### PR DESCRIPTION
**Pull request by gauravtiwari5050**
# Instant Answer for Economic Indicators

**What does your Instant Answer do?**

The Instant Answer provides values for Economic Indicators for  different countries.
The indicators considered are
- Gross Domestic Product
- Annual Growth Rate
- Per Capita Income

All data used in this instant answer comes form World Bank(http://data.worldbank.org)

**What problem does your Instant Answer solve (Why is it better than organic links)?**

It provides exact answer to specific queries. For example if I want know what the GDP of India, I will first have to go to the given organic link and look for the answer. My instant answer gives a specific answer for quries relating to gdp,per capita income and growth rate.

**What is the data source for your Instant Answer? (Provide a link if possible)**
Data Sources 
- Gross Domestic Product http://data.worldbank.org/indicator/NY.GDP.MKTP.CD
- Per Capita Income http://data.worldbank.org/indicator/NY.GNP.PCAP.PP.CD
- Annual Growth Rate http://data.worldbank.org/indicator/NY.GDP.MKTP.KD.ZG

**Why did you choose this data source?**

World Bank is the one of the most accurate and authentic sources of economic data. It is also the most comprehensive data source covering almost all countries.

**Are there any other alternative (better) data sources?**

There are alternative data sources but they are not better than World Bank data. The other data sources are are fragmented or limited to certain geographies.

**What are some example queries that trigger this Instant Answer?**
- what is the per capita income of china
- per capita income, japan
- what is the growth rate of indonesia
- malaysia growth rate
- gdp of india
- sri lanka gdp
- what is the gross domestic product of singapore?

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
- students
- researchers

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**

No

**Which existing Instant Answers will this one supercede/overlap with?**

Supercedes results from Zanran

**Are you having any problems? Do you need our help with anything?**

No

**Where did you hear about DuckDuckHack? (For first time contributors)**

HackerNews

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
Chrome/Mac Screenshot
Chrome/Ubuntu Screenshot
Chrome/Android Screenshot

## Checklist
Please place an 'X' where appropriate.

```
[X] Added metadata and attribution information (https://duck.co/duckduckhack/metadata)
[X] Wrote test file and added to t/ directory (https://duck.co/duckduckhack/goodie_tests)
[X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 8
        [] IE 9
        [] IE 10

    - Windows XP
        [] IE 7
        [] IE 8

    - Mac OSX
        [X] Google Chrome
        [X] Firefox
        [] Opera
        [X] Safari

    - iOS (iPhone)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [] Firefox
        [] Native Browser
        [X] Google Chrome
        [] Opera
```

### CODE REVIEW

**Comment by zekiel** [comment]
@gauravtiwari5050 anything else we can help with? :) just putting this back on your radar if you forgot but, of course, no rush.

**Comment by gauravtiwari5050** [comment]
@zekiel Apologies for the delay. Will try and finish this soon.

**Comment by zekiel** [comment]
No worries @gauravtiwari5050  Looking forward to it!

**Comment by jagtalon** [comment]
@gauravtiwari5050 just ping us if you need any help. :)

**Comment by moollaza** [comment]
ping! Hey @gauravtiwari5050, I'm just checking in to see if you're still interested in finishing this one up? We'd love to see this go live!

File: lib/DDG/Spice/EconomicIndicators.pm [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+topics "economy_and_finance";
+code_url "https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/lib/DDG/Spice/EconomicIndicators.pm";
+attribution github => ["gauravtiwari5050", "Gaurav Tiwari"];
+
+# Triggers
+triggers any => "gdp", "gross domestic product","growth rate","per capita income";
+
+#the world bank api requires date range in years
+#we will use the current year, and three previous years
+spice to => 'http://api.worldbank.org/countries/$1?per_page=10&date='.localtime->add_years(-3)->year.':'.localtime->year.'&format=json';
```
Comment (gauravtiwari5050) [inline lib/DDG/Spice/EconomicIndicators.pm]: For calling the worldbank apis we need a date range, for which the current system time would suffice. We dont neccessarily need the location of the user for this purpose.

File: share/spice/economic_indicators/economic_indicators.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
+(function (env) {
+    "use strict";
+    function getDataFromAPIResponse(api_response){
+        //validate if api response is null
+        if(api_response === null || api_response === undefined){
+            return null;
+        }
+        //the root object of api response must be an array of length 2
+        if (Object.prototype.toString.call( api_response ) !== '[object Array]' || api_response.length != 2){
```
Comment (gauravtiwari5050) [inline share/spice/economic_indicators/economic_indicators.js]: This is done as suggested https://github.com/gauravtiwari5050/zeroclickinfo-spice/commit/8270b5957238fe90716bc98025efe09d9157b4e6

File: share/spice/economic_indicators/economic_indicators.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                    indicatorValue = (indicatorValue/1000000000).toFixed(2) + " Billion USD";
+                    indicatorValue = sprintf("%.3f ",indicatorValue/1000000000);
+                  } else if(indicatorValue/1000000 > 1){
+                    indicatorValue = (indicatorValue/1000000).toFixed(2) + " Million USD";
+                  } else {
+                    indicatorValue = indicatorValue.toFixed(2) + " USD";
+                  }
+                  indicatorData.value = indicatorValue;
+            }
+
```
Comment (gauravtiwari5050) [inline share/spice/economic_indicators/economic_indicators.js]: This is done as suggested.
https://github.com/duckduckgo/zeroclickinfo-spice/commit/5664ba8cb53e6021a91426f32e203ac3835c3f06

File: share/spice/economic_indicators/economic_indicators.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            // if the indicator is per capita income
+            if(indicatorData.indicator.value.match(/per capita/gi)){
+                  indicatorData.value = indicatorData.value + " USD";
+            }
+            //else if indicatos is growth rate
+            else if(indicatorData.indicator.value.match(/growth/gi)){
+                  indicatorData.value = parseFloat(indicatorData.value).toFixed(2) + " %";
+            }
+            // else if the indicator is GDP
+            else if(indicatorData.indicator.value.match(/GDP/gi)){
```
Comment (gauravtiwari5050) [inline share/spice/economic_indicators/economic_indicators.js]: Moved the cleanup code to a normalization block https://github.com/duckduckgo/zeroclickinfo-spice/commit/27e03974d494fe66a91828f533989f39335ac573

File: share/spice/economic_indicators/economic_indicators.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+            // Customize these properties
+            name: "Economic Indicator",
+            data: indicatorData,
+            meta: {
+                sourceName: "WorldBank",
+                sourceUrl: 'http://data.worldbank.org/indicator/' + indicatorData.indicator.id
+            },
+            templates: {
+                group: 'base',
```
Comment (gauravtiwari5050) [inline share/spice/economic_indicators/economic_indicators.js]: Done as suggested. https://github.com/duckduckgo/zeroclickinfo-spice/commit/602c70f94b7c10f8efa25d1e957282cfd852792c

**Comment by gauravtiwari5050** [comment]
@moollaza Apologies for the delay on this one.
I have made the changes as suggested in the review.
Could you please have a look again ?

Thanks,
Gaurav

File: lib/DDG/Spice/EconomicIndicators.pm (hunk 1 above)
Comment (moollaza) [inline lib/DDG/Spice/EconomicIndicators.pm]: @gauravtiwari5050 sorry to clarify I meant that by using the `$loc` variable, you can get the timezone of the user to calculate the date in their timezone. `localtime` refers to the server's time zone which isn't necessarily the same as the users. However, seeing as you're using it to calculate the year, this is be fine :+1:

**Comment by moollaza** [comment]
@gauravtiwari5050 thanks a lot for fixing everything up! We'll give this a review shortly

**Comment by jagtalon** [comment]
@gauravtiwari5050 nice! It looks pretty much like the rainfall IA:

screen shot 2015-02-06 at 4 01 57 am
screen shot 2015-02-06 at 4 02 02 am

There's a bit of a spacing issue, but I think we can do away without a custom handlebars template here and just follow what rainfall is using. Check it out! I think it will be a bit simpler. :) CC @abeyang 

https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/share/spice/rainfall/rainfall.js

A similar thing is being used in SalesTax, although it's not live yet: https://github.com/duckduckgo/zeroclickinfo-spice/pull/1474/files

Thank you!

**Comment by gauravtiwari5050** [comment]
@jagtalon I have removed the handlebars template and the css file.
These are the updated screenshots
updated_economic_indicator_screenshot

screenshot_2015-02-06-23-39-01

**Comment by moollaza** [comment]
@gauravtiwari5050 looking good! The only thing I noticed is that the `20990 USD` should probably have a separator after the `20`, i.e. `20,990 USD`

@jagtalon can you throw this on a server ?

**Comment by gauravtiwari5050** [comment]
@moollaza @jagtalon I have added comma seperation as suggested.
brazil_pci

**Comment by jagtalon** [comment]
@abeyang @moollaza it's up https://ddh9.duckduckgo.com/?q=gdp+of+india&ia=economicindicator. I think I'd like it to be "current USD" instead of "current US$" but that's just me.

**Comment by jagtalon** [comment]
@gauravtiwari5050 can the "More at" link to http://data.worldbank.org/country/india instead of http://data.worldbank.org/indicator/NY.GDP.MKTP.CD?

**Comment by abeyang** [comment]
@gauravtiwari5050 can we have it like so:

> **1.88 Trillion USD**
> India - GDP (2013)

Please note the hyphen and removing "current US$" altogether -- don't think it's necessary. :smile:

File: share/spice/economic_indicators/economic_indicators.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
+(function (env) {
+    "use strict";
+    function getDataFromAPIResponse(api_response){
+        //validate if api response is null
+        if(api_response === null || api_response === undefined){
+            return null;
```
Comment (jagtalon) [inline share/spice/economic_indicators/economic_indicators.js]: We need to return `Spice.failed` like in https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/share/spice/alternative_to/alternative_to.js#L7

Comment (jagtalon) [inline share/spice/economic_indicators/economic_indicators.js]: Ack, was reading the wrong part. This is good. :)

**Comment by moollaza** [comment]
@gauravtiwari5050 ping! Are you still interested in finishing this up?

**Comment by gauravtiwari5050** [comment]
@moollaza Apologies for the delay
I have made the changes per @jagtalon and @abeyang suggestions.

**Comment by moollaza** [comment]
@gauravtiwari5050 thanks a lot! Everything LGTM :+1:

**Comment by moollaza** [comment]
@gauravtiwari5050 Thanks again for taking the time to contribute! We really appreciate it. Apologies that this took so long, We're constantly trying to improve our process so IA's can move along quicker and smoother. Feel free to stick around and comment/help on other PR's & Issues or even submit more Instant Answers! I'll try my hardest to make sure they don't take nearly as long this time :)

**Comment by gauravtiwari5050** [comment]
@moollaza Thanks a lot!

**Comment by gauravtiwari5050** [comment]
@moollaza duckduckgo.com is not using this instant answer as of now, Do i need to do something else to get this on production duckduckgo ?

**Comment by moollaza** [comment]
@gauravtiwari5050 nope, it will be live shortly. The deploy process isn't immediate -- I'll ping you here when it goes live :)

**Comment by moollaza** [comment]
Sorry, to clarify it would usually be up a little sooner than this -- we ran into some internal bugs that slowed the release but it's set to go live with the next deploy.

**Comment by moollaza** [comment]
@gauravtiwari5050 congrats! It's live: https://duckduckgo.com/?q=gdp+of+india&ia=economicindicator

Sorry for the delay, this has been live for about a week now!

Thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's & Issues or even submit more Instant Answers!

**Pull request merged**